### PR TITLE
fix: Add ENV variable to configure GRPC Keep Alive Time (#15656)

### LINKED ISSUE #15656: ENHANCE_YOUR_CALM error for plugins that take >45 seconds
Checklist:

* [X ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ X] I've included steps to reproduce the bug.
* [ X] I've pasted the output of `argocd version`.

**Describe the bug**

When creating an application that runs a custom plugin that runs for over ~45 seconds, the user is returned an `ENHANCE_YOUR_CALM` error.

This happens on occasion for plugins > 45s and consistently for plugins > 80s.

**To Reproduce**

Create a CMP that runs for a long time (e.g. run a script that includes a `sleep 90`). 

Run:
```
argocd app create test-app --file create.json --grpc-web
```

where `create.json` is the definition of the app that uses the long-running CMP.

**Expected behavior**

App is created with no error.

**Screenshots**

N/A

**Version**

```shell
argocd: v2.8.4+c279299.dirty
  BuildDate: 2023-09-13T22:00:14Z
  GitCommit: c27929928104dc37b937764baf65f38b78930e59
  GitTreeState: dirty
  GoVersion: go1.21.1
  Compiler: gc
  Platform: darwin/amd64

```

**Logs**

```
2023/09/23 16:45:34 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
FATA[0080] rpc error: code = Unavailable desc = closing transport due to: connection error: desc = "error reading from server: EOF", received prior goaway: code: ENHANCE_YOUR_CALM, debug data: "too_many_pings"
```

### PR DESCRIPTION
**Pull request by BhavikaSharma**
Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [X] I have added a brief description of why this PR is necessary and/or what this PR solves.

Description:
When creating an application that runs a custom plugin that runs for over ~45 seconds, the user is returned an `ENHANCE_YOUR_CALM` error. The CMP timeout is a configurable value with a default of 60s. Multiple `argocd app` commands run the CMP (e.g. as validation). Therefore, the `GRPCKeepAliveEnforcementMinimum` should be configurable.

Closes #15656

### CODE REVIEW

**Comment by BhavikaSharma** [comment]
There seems to be flakiness in the e2e tests. A different test seems to be failing in each run despite no changes, e.g.

Passing: https://github.com/argoproj/argo-cd/actions/runs/6401735616/job/17377379600
Failing: https://github.com/argoproj/argo-cd/actions/runs/6409855823/job/17403620905

Passing: https://github.com/argoproj/argo-cd/actions/runs/6401032458/job/17375509894
Failing: https://github.com/argoproj/argo-cd/actions/runs/6401735616/job/17377289561

Rerunning the pipeline as described [here](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/#why-does-the-test-or-e2e-steps-fail).

**Comment by renperez-cpi** [comment]
@BhavikaSharma any update on this?

**Comment by jcogilvie** [comment]
The most recent failure seems to indicate that upstream redis has changed.  Is this a problem for this PR?

**Comment by crenshaw-dev** [comment]
@jcogilvie merged upstream, should fix the CI failure.

**Comment by BhavikaSharma** [comment]
@crenshaw-dev , thanks for merging in the upstream. We seem to have a green build.

**Comment by renperez-cpi** [comment]
@BhavikaSharma is this going to be merge soon? argo seems to be getting worst.
```
2023/10/16 19:29:00 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
2023/10/16 19:41:12 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
2023/10/16 19:50:21 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
2023/10/16 19:54:25 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
2023/10/16 20:06:37 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
2023/10/16 20:16:47 ERROR: [transport] Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings".
time="2023-10-16T20:31:25Z" level=fatal msg="rpc error: code = Unknown desc = POST https://***:4[43](https://github.com/ChargePoint/payment-services/actions/runs/6535831960/job/17753819951#step:9:44)/application.ApplicationService/Get failed with status code 504"
Error: Process completed with exit code 20.
```

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
As https://github.com/argoproj/argo-cd/pull/15708 was merged, can you add pkg/apiclient/grpcproxy.go to the change as well?

**Comment by BhavikaSharma** [comment]
I have added `pkg/apiclient/grpcproxy.go` to the change as well.

File: docs/user-guide/environment-variables.md [hunk 1]
```diff
@@ -12,3 +12,4 @@ The following environment variables can be used with `argocd` CLI:
 | `ARGOCD_APPLICATION_CONTROLLER_NAME` | the Argo CD Application Controller name (default "argocd-application-controller")                                                                                                                         |
 | `ARGOCD_REDIS_NAME`                  | the Argo CD Redis name (default "argocd-redis")                                                                                                                                                           |
 | `ARGOCD_REDIS_HAPROXY_NAME`          | the Argo CD Redis HA Proxy name (default "argocd-redis-ha-haproxy")                                                                                                                                       |
+| `ARGOCD_GRPC_KEEP_ALIVE_MIN`         | defines the GRPCKeepAliveEnforcementMinimum, used in the grpc.KeepaliveEnforcementPolicy. Expects a "Duration" format (e.g. `10s`).                                                                       |
```
Comment (ishitasequeira) [inline docs/user-guide/environment-variables.md:15]: Let's mention the default value. 
```suggestion
| `ARGOCD_GRPC_KEEP_ALIVE_MIN`         | defines the GRPCKeepAliveEnforcementMinimum, used in the grpc.KeepaliveEnforcementPolicy. Expects a "Duration" format (default `10s`).                                                                       |
```

**Review by ishitasequeira — APPROVED** [review]
One minor suggestion. Otherwise LGTM!!

**Comment by BhavikaSharma** [comment]
@ishitasequeira thanks for the approval! I have added the change suggested above and merged in the latest changes from main with a passing pipeline. If it looks good, would you be able to help merge? Thanks 😄 .

**Review by ishitasequeira — APPROVED** [review]

**Pull request merged**